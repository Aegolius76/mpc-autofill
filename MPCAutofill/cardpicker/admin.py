from django.contrib import admin

from .models import Card, Cardback, DFCPair, Source, Token


# Register your models here.
@admin.register(Card)
class AdminCard(admin.ModelAdmin):
    list_display = ("name", "source", "dpi", "date")


@admin.register(Cardback)
class AdminCardback(admin.ModelAdmin):
    list_display = ("name", "source", "dpi", "date")


@admin.register(Token)
class AdminToken(admin.ModelAdmin):
    list_display = ("name", "source", "dpi", "date")


@admin.register(DFCPair)
class AdminDFCPair(admin.ModelAdmin):
    list_display = ("front", "back")


@admin.register(Source)
class AdminSource(admin.ModelAdmin):
    list_display = ("id", "drive_id", "contribution", "description")

    def contribution(self, obj):
        qty_all, qty_cards, qty_cardbacks, qty_tokens, avgdpi = obj.count()

        return "{} images - {} cards, {} cardbacks, and {} tokens @ {} DPI on average".format(
            qty_all, qty_cards, qty_cardbacks, qty_tokens, avgdpi
        )
