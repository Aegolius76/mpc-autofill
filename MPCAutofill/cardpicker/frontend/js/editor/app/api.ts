// TODO: set up the below API calls to use best practices with handling all cases `fetch` can return
import Cookies from "js-cookie";
import { CSRFToken } from "../common/constants";
import {
  SearchSettingsState,
  CardDocuments,
  SearchResults,
  SourceDocuments,
  SearchQuery,
  DFCPairs,
} from "../common/types";

export async function APIGetCards(
  identifiersToSearch: Set<string>
): Promise<CardDocuments> {
  const rawResponse = await fetch("/2/getCards/", {
    method: "POST",
    body: JSON.stringify({
      card_identifiers: Array.from(identifiersToSearch),
    }),
    credentials: "same-origin",
    headers: {
      "X-CSRFToken": Cookies.get(CSRFToken),
    },
  });
  const content = await rawResponse.json();
  return content.results;
}

export async function APIGetCardbacks(): Promise<Array<string>> {
  const rawResponse = await fetch("/2/getCardbacks/", {
    method: "GET",
    credentials: "same-origin",
    headers: {
      "X-CSRFToken": Cookies.get(CSRFToken),
    },
  });
  const content = await rawResponse.json();
  return content.cardbacks;
}

export async function APISearch(
  searchSettings: SearchSettingsState,
  queriesToSearch: Array<SearchQuery>
): Promise<SearchResults> {
  const rawResponse = await fetch("/2/search/", {
    method: "POST",
    body: JSON.stringify({
      searchSettings,
      queries: Array.from(queriesToSearch),
    }),
    credentials: "same-origin",
    headers: {
      "X-CSRFToken": Cookies.get(CSRFToken),
    },
  });
  const content = await rawResponse.json();
  return content.results;
}

export async function APIGetSources(): Promise<SourceDocuments> {
  const rawResponse = await fetch("/2/getSources/", {
    method: "GET",
    credentials: "same-origin",
    headers: {
      "X-CSRFToken": Cookies.get(CSRFToken),
    },
  });
  const content = await rawResponse.json();
  return content.results;
}

export async function APIGetImportSites() {
  const rawResponse = await fetch("/2/getImportSites", {
    method: "GET",
    credentials: "same-origin",
    headers: {
      "X-CSRFToken": Cookies.get(CSRFToken),
    },
  });
  const content = await rawResponse.json();
  return content.import_sites;
}

export async function APIQueryImportSite(url: string): Promise<string> {
  const rawResponse = await fetch("/2/queryImportSite/", {
    method: "POST",
    body: JSON.stringify({ url }),
    credentials: "same-origin",
    headers: {
      "X-CSRFToken": Cookies.get(CSRFToken),
    },
  });
  const content = await rawResponse.json();
  return content.cards;
}

export async function APIGetDFCPairs(): Promise<DFCPairs> {
  const rawResponse = await fetch("/2/getDFCPairs/", {
    method: "GET",
    credentials: "same-origin",
    headers: {
      "X-CSRFToken": Cookies.get(CSRFToken),
    },
  });
  const content = await rawResponse.json();
  return content.dfc_pairs;
}
