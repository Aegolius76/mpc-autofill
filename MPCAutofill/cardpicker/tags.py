"""
Tags for cards.
"""

import functools
import re
from typing import Optional

from aenum import MultiValueEnum


class Tag(str, MultiValueEnum):
    ALT_ART = "ALT ART", "ALTART", "ALT"
    EXTENDED = "EXTENDED"
    FULL_ART = "FULL ART", "FULLART", "FULL"
    NSFW = "NSFW"

    @staticmethod
    def extract_name_and_tags(name: Optional[str]) -> tuple[str, list[str]]:
        """
        This function unpacks a folder or image name which contains a name component and some number of tags
        into its constituents.
        Tags are wrapped in either [square brackets] or (parentheses), and any combination of [] and () can be used
        within a single name.
        """

        if not name:
            return "", []

        tag_set = set()  # Use set to not have duplicate tags
        tag_parts = re.findall(r"\(([^\(\)]+)\)|\[([^\[\]]+)\]", name)  # Get content of () and []
        cleaned_parts = list(map(lambda x: x[0] if len(x[0]) != 0 else x[1], tag_parts))
        name_with_no_tags = functools.reduce(
            lambda mutated_name, tag: mutated_name.replace(f"[{tag}]", "").replace(f"({tag})", "").strip(),
            [name, *cleaned_parts],
        )
        for tag_part in cleaned_parts:
            raw_tags = tag_part.split(",")
            for raw_tag in raw_tags:
                try:
                    tag_set.add(Tag(raw_tag.strip().upper()))
                except ValueError:
                    pass  # Unknown tag

        return name_with_no_tags, [x.value for x in tag_set]
