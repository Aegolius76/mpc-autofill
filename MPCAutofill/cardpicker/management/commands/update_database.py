from django.core.management.base import BaseCommand
from django.db import transaction, connection
from django.core import management
from cardpicker.models import Card, Cardback, Token, Source
import pickle
from googleapiclient.errors import HttpError
from googleapiclient.discovery import build
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request
import time
import os
from concurrent.futures import ThreadPoolExecutor
from functools import partial
from tqdm import tqdm
from to_searchable import to_searchable
import colorama
import csv
from math import floor


def int_mean(input_list):
    # cleaned_list = [x for x in input_list if x.isdigit()]
    if len(input_list) == 0:
        return 0
    try:
        return int(sum(input_list)/len(input_list))
    except ZeroDivisionError:
        return 0


# If modifying these scopes, delete the file token.pickle.
SCOPES = ['https://www.googleapis.com/auth/drive.metadata.readonly',
          'https://www.googleapis.com/auth/drive.readonly']

global SOURCES
SOURCES = {}

q_cards = []
q_cardbacks = []
q_tokens = []


# read CSV file for drive data
with open("drives.csv", newline='') as csvfile:
    drivesreader = csv.DictReader(csvfile, delimiter=",")
    for row in drivesreader:
        SOURCES[row["key"]] = {
            "qty_cards": 0,
            "qty_cardbacks": 0,
            "qty_tokens": 0,
            "username": row["username"],
            "reddit": row["reddit"],
            "drivelink": row["drivelink"],
            "description": row["description"],
            "drivename": row["drivename"]
        }

SOURCES["Unknown"] = {
    "quantity": 0,
    "username": "",
    "reddit": "",
    "drivelink": "",
    "description": "",
    "drivename": ""
}

global OWNERS
OWNERS = {SOURCES[x]["drivename"]: x for x in SOURCES.keys()}

DPI_HEIGHT_RATIO = 300/1100  # 300 DPI for image of vertical resolution 1100 pixels


def fill_tables(service):
    # Call to google drive API
    results = service.files().list(
        q="mimeType='application/vnd.google-apps.folder' and sharedWithMe=true",
        fields="files(id, name, parents, driveId)",
        pageSize=1000
    ).execute()

    folders = results.get('files', [])
    print("Folders found: {}".format(len(folders)))

    for folder in folders:
        search_folder(service, folder)
    # search_folder(service, folders[22])

    # Commit changes for each table in bulk
    print("Committing cards to tables")
    with connection.cursor() as cursor:
        cursor.execute('DELETE FROM cardpicker_card')
        cursor.execute('DELETE FROM cardpicker_cardback')
        cursor.execute('DELETE FROM cardpicker_token')

    Card.objects.bulk_create(
        list([Card(
            id=x[0],
            name=x[1],
            priority=x[2],
            source=x[3],
            dpi=x[4],
            searchq=x[5],
            thumbpath=x[6]
        ) for x in q_cards])
    )

    Cardback.objects.bulk_create(
        list([Cardback(
            id=x[0],
            name=x[1],
            priority=x[2],
            source=x[3],
            dpi=x[4],
            searchq=x[5],
            thumbpath=x[6]
        ) for x in q_cardbacks])
    )

    Token.objects.bulk_create(
        list([Token(
            id=x[0],
            name=x[1],
            priority=x[2],
            source=x[3],
            dpi=x[4],
            searchq=x[5],
            thumbpath=x[6]
        ) for x in q_tokens])
    )
    print("Done with cards")


def search_folder(service, folder):
    # folders to skip
    ignoredFolders = [
        '3x5 Size',
        '3.5x5 Size',
        '11. Planechase',
        '!Chili_Axe Card Backs',
        '!Card Backs',
        '[EXTRA] - Card back',
        '[Update 6/5/18] Legendary Walkers',
        '[Update: 6/10/18] Redirect & Misc Errata',
        'Cubes',
        'X. Art & Misc Stuff',
    ]

    print("Searching drive: {}".format(folder['name']))

    folderList = [folder]
    imageList = []
    folderDict = {}
    parentDict = {}

    while len(folderList) > 0:
        # Add all folders within the current folder to the folder list
        # Add all images within the current folder to the images list
        # Remove the current folder from the folder list
        # The next current folder is the 1st element in the folder list
        time.sleep(0.1)
        currFolder = folderList[0]
        # Skip some folders as specified
        acceptable = all(x not in currFolder['name'] for x in ignoredFolders)
        if acceptable:
            print("Searching: {}".format(currFolder['name']))
            folderDict[currFolder['id']] = currFolder['name']
            try:
                parentDict[currFolder['id']] = folderDict[currFolder['parents'][0]]
            except KeyError:
                parentDict[currFolder['id']] = ""

            # Search for folders within the current folder
            while True:
                try:
                    results = service.files().list(
                        q="mimeType='application/vnd.google-apps.folder' and "
                          "'{}' in parents".format(currFolder['id']),
                        fields="files(id, name, parents)",
                        pageSize=500
                    ).execute()
                    folderList += results.get('files', [])
                    break
                except HttpError:
                    pass

            # Search for images with paging - probably not necessary for folders
            page_token = None
            while True:
                # Search for all images within this folder
                time.sleep(0.1)
                try:
                    results = service.files().list(
                        q="(mimeType contains 'image/png' or "
                          "mimeType contains 'image/jpg' or "
                          "mimeType contains 'image/jpeg') and "
                          "'{}' in parents".format(currFolder['id']),
                        fields="nextPageToken, files("
                               "id, name, trashed, size, properties, parents, modifiedTime, imageMediaMetadata, owners"
                               ")",
                        pageSize=500,
                        pageToken=page_token
                    ).execute()
                except HttpError:
                    # TODO: Not pass?
                    pass

                if len(results['files']) <= 0:
                    break

                print("Found {} image(s)".format(len(results.get('files', []))))

                # Append the retrieved images to the image list
                imageList += results.get('files', [])

                page_token = results.get('nextPageToken', None)
                if page_token is None:
                    break
        folderList.remove(currFolder)

    print("Number of images found: {}".format(len(imageList)))
    print("Folder dict:")
    for key in folderDict:
        print("{}: {}".format(key, folderDict[key]))

    # add the retrieved cards to the database, parallelised by 20 for speed
    # TODO: Is 20 the right number here?
    with tqdm(total=len(imageList), desc="Collecting card results") as bar, ThreadPoolExecutor(max_workers=20) as pool:
        for _ in pool.map(partial(add_card, folderDict, parentDict, folder), imageList):
            bar.update(1)

    print("Finished crawling {}".format(folder['name']))
    print("")


def add_card(folderDict, parentDict, folder, item):
    try:
        # file is valid when it's not trashed and filesize does not exceed 30 MB
        valid = not item['trashed'] and int(item['size']) < 30000000
        if not valid:
            print("Can't index this card: {}".format(item))
    except KeyError:
        valid = True

    if valid:
        # strip the extension off of the item name to retrieve the card name
        try:
            [cardname, extension] = item['name'].rsplit('.', 1)
        except ValueError:
            print("nani {}".format(item['name']))
            return

        img_type = "normal"
        folderName = folderDict[item['parents'][0]]
        parentName = parentDict[item['parents'][0]]

        owner = item['owners'][0]['displayName']

        scryfall = False
        priority = 2
        if "Retro Cube" in parentName:
            priority = 0
        if ")" in cardname:
            priority = 1
        source = "Unknown"
        if folder['name'] == "Chilli_Axe's MPC Proxies":
            source = "Chilli_Axe"
            if folderName == "12. Cardbacks":
                if "Black Lotus" in item['name']:
                    priority += 10
                img_type = "cardback"
                priority += 5

        elif folder['name'] == "nofacej MPC Card Backs":
            img_type = "cardback"
            source = "nofacej"

        elif owner in OWNERS:
            source = OWNERS[owner]

        elif folder['name'] == "MPC Scryfall Scans":
            source = "berndt_toast83/" + folderName
            scryfall = True

        if "Basic" in folderName:
            priority += 5

        if "Token" in folderName:
            img_type = "token"

        # Store the image's static URL
        static_url = "https://drive.google.com/thumbnail?sz=w400-h400&id=" + item['id']

        # Calculate source image DPI, rounded to tens
        dpi = 10 * round(int(item['imageMediaMetadata']['height']) * DPI_HEIGHT_RATIO / 10)

        # Return card info so we can insert into database, in the correct list
        card_info = (item['id'], cardname, priority, source, dpi, to_searchable(cardname), extension)

        if img_type == "cardback":
            q_cardbacks.append(card_info)
            if scryfall:
                SOURCES["berndt_toast83"]["qty_cardbacks"] += 1
            else:
                SOURCES[source]["qty_cardbacks"] += 1
        elif img_type == "token":
            q_tokens.append(card_info)
            if scryfall:
                SOURCES["berndt_toast83"]["qty_tokens"] += 1
            else:
                SOURCES[source]["qty_tokens"] += 1
        else:
            q_cards.append(card_info)
            if scryfall:
                SOURCES["berndt_toast83"]["qty_cards"] += 1
            else:
                SOURCES[source]["qty_cards"] += 1


def add_sources():
    source_ids = list(SOURCES.keys())
    source_ids.remove("Unknown")

    print("Committing sources to tables")
    Source.objects.all().delete()

    with transaction.atomic():
        Source.objects.all().delete()
        for source_id in source_ids:
            source_dpis = []
            for model in [Card, Cardback, Token]:
                source_dpis.extend(list(model.objects.filter(source__startswith=source_id).values_list('dpi', flat=True)))

            Source.objects.create(
                id=source_id,
                qty_cards=SOURCES[source_id]['qty_cards'],
                qty_cardbacks=SOURCES[source_id]['qty_cardbacks'],
                qty_tokens=SOURCES[source_id]['qty_tokens'],
                username=SOURCES[source_id]['username'],
                reddit=SOURCES[source_id]['reddit'],
                drivelink=SOURCES[source_id]['drivelink'],
                description=SOURCES[source_id]['description'],
                avgdpi=int_mean(source_dpis)
            )

    print("Done with sources")


def login():
    creds = None
    # The file token.pickle stores the user's access and refresh tokens, and is
    # created automatically when the authorization flow completes for the first
    # time.
    if os.path.exists('token.pickle'):
        with open('token.pickle', 'rb') as token:
            creds = pickle.load(token)
    # If there are no (valid) credentials available, let the user log in.
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file(
                'credentials.json', SCOPES)
            creds = flow.run_local_server(port=0)
        # Save the credentials for the next run
        with open('token.pickle', 'wb') as token:
            pickle.dump(creds, token)

    return build('drive', 'v3', credentials=creds)


class Command(BaseCommand):
    def handle(self, *args, **options):
        service = login()
        t = time.time()
        fill_tables(service)
        add_sources()
        for source in SOURCES:
            print(SOURCES[source])
        t_final = time.time()
        mins = floor((t_final - t) / 60)
        secs = int((t_final - t) - mins * 60)
        print("Elapsed time: {} minutes and {} seconds.\n".format(mins, secs))

        print("Rebuilding search index")
        management.call_command('search_index', '--rebuild', '-f')
