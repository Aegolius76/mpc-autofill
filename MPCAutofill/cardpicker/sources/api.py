import datetime as dt
import functools
import os
import re
import threading
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Optional

import pycountry
import ratelimit
from googleapiclient.discovery import Resource, build
from googleapiclient.errors import HttpError
from oauth2client.service_account import ServiceAccountCredentials

from cardpicker.constants import DEFAULT_LANGUAGE
from cardpicker.tags import Tag

thread_local = threading.local()  # Should only be called once per thread


@dataclass
class Folder:
    id: str
    name: str
    parent: Optional["Folder"]

    @functools.cached_property
    def top_level_folder(self) -> "Folder":
        if self.parent is None:
            return self
        return self.parent.top_level_folder

    @functools.cached_property
    def full_path(self) -> str:
        _, name, _ = self.unpacked_name
        if self.parent is None:
            return name
        return f"{self.parent.full_path} / {name}"

    @functools.cached_property
    def unpacked_name(self) -> tuple[pycountry.Languages, str, set[str]]:
        """
        The folder's name is unpacked according to the below schema. For example, consider `<EN> Cards [NSFW]`:
             <EN>              Cards         [NSFW]
        └─ language ──┘ └─ folder name ──┘ └─ tags ──┘
        """

        folder_name_results = re.compile(r"^(?:<(.+)> )?(.*?)$").search(self.name)
        assert folder_name_results is not None
        language_code, name = folder_name_results.groups()
        language = (
            (pycountry.languages.get(alpha_2=language_code) or DEFAULT_LANGUAGE)
            if language_code is not None
            else DEFAULT_LANGUAGE
        )
        name_with_no_tags, tags = Tag.extract_name_and_tags(name)
        return language, name_with_no_tags, tags

    @functools.cached_property
    def language(self) -> pycountry.Languages:
        language, _, _ = self.unpacked_name
        if self.parent is None:
            return language
        return language if language != DEFAULT_LANGUAGE else self.parent.language

    @functools.cached_property
    def tags(self) -> set[str]:
        _, _, tags = self.unpacked_name
        if self.parent is None:
            return tags
        return self.parent.tags | tags


@dataclass
class Image:
    id: str
    name: str
    size: int
    created_time: dt.datetime
    height: int
    folder: Folder

    @functools.cached_property
    def unpacked_name(self) -> tuple[pycountry.Languages, str, set[str], str]:
        """
        The image's name is unpacked according to the below schema. For example, consider `<EN> Opt [NSFW].png`:
             <EN>             opt          [NSFW]   .      png
        └─ language ──┘ └─ card name ──┘ └─ tags ──┘ └─ extension ──┘
        """

        assert self.name, "File name is empty string"
        image_name_results = re.compile(r"^(?:<(.+)> )?(.*?)(?:\.(.*?))?$").search(self.name)
        assert image_name_results is not None
        language_code, name, extension = image_name_results.groups()
        language = (
            (pycountry.languages.get(alpha_2=language_code) or DEFAULT_LANGUAGE)
            if language_code is not None
            else DEFAULT_LANGUAGE
        )
        assert extension is not None, "File name has no extension"
        name_with_no_tags, tags = Tag.extract_name_and_tags(name)
        return language, name, tags, extension

    @functools.cached_property
    def language(self) -> pycountry.Languages:
        language, _, _, _ = self.unpacked_name
        return language if language != DEFAULT_LANGUAGE else self.folder.language

    @functools.cached_property
    def tags(self) -> set[str]:
        _, _, image_tags, _ = self.unpacked_name
        return image_tags | self.folder.tags


# region google drive API
# Google Drive API usage limits reference: https://developers.google.com/drive/api/guides/limits


# If modifying these scopes, delete the file token.pickle.
SCOPES = [
    "https://www.googleapis.com/auth/drive.metadata.readonly",
    "https://www.googleapis.com/auth/drive.readonly",
]

SERVICE_ACC_FILENAME = "client_secrets.json"


def find_or_create_google_drive_service() -> Resource:
    if (service := getattr(thread_local, "google_drive_service", None)) is None:
        creds = ServiceAccountCredentials.from_json_keyfile_name(
            str(Path(os.path.abspath(__file__)).parent.parent.parent / SERVICE_ACC_FILENAME), scopes=SCOPES
        )
        service = build("drive", "v3", credentials=creds)
        thread_local.google_drive_service = service
    return service


@ratelimit.sleep_and_retry  # type: ignore  # `ratelimit` does not implement decorator typing correctly
@ratelimit.limits(calls=20_000, period=100)  # type: ignore  # `ratelimit` does not implement decorator typing correctly
def execute_google_drive_api_call(service: Resource) -> Optional[dict[str, Any]]:
    try:
        return service.execute()
    except HttpError:
        return {}


# endregion
