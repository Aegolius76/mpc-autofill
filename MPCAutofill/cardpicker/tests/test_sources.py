import datetime as dt

import pytest

from cardpicker.models import Card
from cardpicker.sources.api import Folder, Image
from cardpicker.sources.update_database import update_database
from cardpicker.tags import read_tags_in_database


class TestAPI:
    # region constants

    DEFAULT_DATE = dt.datetime(2023, 1, 1)

    FOLDER_A = Folder(id="a", name="Folder A", parent=None)
    FOLDER_B = Folder(id="b", name="Folder B", parent=FOLDER_A)
    FOLDER_C = Folder(id="c", name="Folder C [NSFW]", parent=FOLDER_B)
    FOLDER_D = Folder(id="d", name="Folder D [Tag in data]", parent=FOLDER_B)
    FOLDER_E = Folder(id="e", name="Folder E [tagindata]", parent=FOLDER_B)  # refers to the tag's alias
    FOLDER_F = Folder(id="f", name="Folder F [tagindata]", parent=FOLDER_B)  # refers to the tag's alias
    FOLDER_G = Folder(id="g", name="Folder G [Tag in Data] (Some more words)", parent=FOLDER_B)
    FOLDER_H = Folder(id="h", name="Folder H [Tag in Data, Some more words]", parent=None)
    FOLDER_X = Folder(id="x", name="Folder X [NSFW, Extended, Full Art]", parent=None)
    FOLDER_Y = Folder(id="y", name="Folder Y [full art, Invalid Tag]", parent=None)
    FOLDER_Z = Folder(id="z", name="Folder z [Full Art", parent=None)
    FOLDER_FRENCH = Folder(id="french", name="<FR> Folder", parent=None)

    IMAGE_A = Image(id="a", name="Image A.png", size=1, created_time=DEFAULT_DATE, height=1, folder=FOLDER_A)
    IMAGE_B = Image(id="b", name="Image B [NSFW].png", size=1, created_time=DEFAULT_DATE, height=1, folder=FOLDER_A)
    IMAGE_C = Image(id="b", name="Image C.png", size=1, created_time=DEFAULT_DATE, height=1, folder=FOLDER_C)
    IMAGE_D = Image(
        id="b", name="Image D [NSFW, full art].png", size=1, created_time=DEFAULT_DATE, height=1, folder=FOLDER_C
    )
    IMAGE_E = Image(
        id="e", name="Image E [invalid tag.png", size=1, created_time=DEFAULT_DATE, height=1, folder=FOLDER_A
    )
    IMAGE_F = Image(
        id="F", name="Image F [NSFW, tag in data].png", size=1, created_time=DEFAULT_DATE, height=1, folder=FOLDER_A
    )
    IMAGE_G = Image(
        id="G", name="Image G [NSFW] (John Doe).png", size=1, created_time=DEFAULT_DATE, height=1, folder=FOLDER_A
    )
    IMAGE_FRENCH = Image(
        id="french", name="French.png", size=1, created_time=DEFAULT_DATE, height=1, folder=FOLDER_FRENCH
    )
    IMAGE_ENGLISH = Image(
        id="english", name="<EN> English.png", size=1, created_time=DEFAULT_DATE, height=1, folder=FOLDER_FRENCH
    )
    IMAGE_NSFW = Image(id="nsfw", name="NSFW [NSFW].png", size=1, created_time=DEFAULT_DATE, height=1, folder=FOLDER_A)
    IMAGE_DOUBLE_NSFW = Image(
        id="double nsfw", name="NSFW (NSFW) [NSFW].png", size=1, created_time=DEFAULT_DATE, height=1, folder=FOLDER_A
    )

    # endregion

    # region tests

    @pytest.mark.parametrize(
        "folder, full_path",
        [(FOLDER_A, "Folder A"), (FOLDER_B, "Folder A / Folder B"), (FOLDER_C, "Folder A / Folder B / Folder C")],
    )
    def test_folder_full_path(self, folder, full_path):
        assert folder.full_path == full_path

    @pytest.mark.parametrize(
        "folder, expected_language",
        [
            (FOLDER_A, None),
            (FOLDER_FRENCH, "FR"),
        ],
    )
    def test_folder_language(self, folder, expected_language):
        if expected_language is None:
            assert folder.language is None
        else:
            assert folder.language.alpha_2.lower() == expected_language.lower()

    @pytest.mark.parametrize(
        "folder, expected_tags",
        [
            (FOLDER_A, set()),
            (FOLDER_B, set()),
            (FOLDER_C, {"NSFW"}),
            (FOLDER_D, {"Tag In Data"}),
            (FOLDER_E, {"Tag In Data"}),
            (FOLDER_X, {"NSFW", "Extended", "Full Art"}),
            (FOLDER_Y, {"Full Art"}),
            (FOLDER_Z, set()),
        ],
    )
    def test_folder_tags(self, django_settings, tags, folder, expected_tags):
        read_tags_in_database()
        assert folder.tags == expected_tags

    @pytest.mark.parametrize(
        "folder, expected_name",
        [
            (FOLDER_A, "Folder A"),
            (FOLDER_B, "Folder B"),
            (FOLDER_C, "Folder C"),
            (FOLDER_G, "Folder G (Some more words)"),
            (FOLDER_H, "Folder H [Some more words]"),
        ],
    )
    def test_folder_name(self, django_settings, tags, folder, expected_name):
        read_tags_in_database()
        _, name, _ = folder.unpacked_name
        assert name == expected_name

    @pytest.mark.parametrize(
        "image, expected_language",
        [
            (IMAGE_A, None),
            (IMAGE_FRENCH, "FR"),
            (IMAGE_ENGLISH, "EN"),
        ],
    )
    def test_image_language(self, image, expected_language):
        if expected_language is None:
            assert image.language is None
        else:
            assert image.language.alpha_2.lower() == expected_language.lower()

    @pytest.mark.parametrize(
        "image, expected_tags",
        [
            (IMAGE_A, set()),
            (IMAGE_B, {"NSFW"}),
            (IMAGE_C, {"NSFW"}),
            (IMAGE_D, {"NSFW", "Full Art"}),
            (IMAGE_E, set()),
            (IMAGE_F, {"NSFW", "Tag In Data"}),
        ],
    )
    def test_image_tags(self, django_settings, tags, image, expected_tags):
        read_tags_in_database()
        assert image.tags == expected_tags

    @pytest.mark.parametrize(
        "image, expected_name",
        [
            (IMAGE_A, "Image A"),
            (IMAGE_B, "Image B"),
            (IMAGE_C, "Image C"),
            (IMAGE_D, "Image D"),
            (IMAGE_E, "Image E [invalid tag"),
            (IMAGE_F, "Image F"),
            (IMAGE_G, "Image G (John Doe)"),
            (IMAGE_NSFW, "NSFW"),
            (IMAGE_DOUBLE_NSFW, "NSFW"),
        ],
    )
    def test_image_name(self, django_settings, tags, image, expected_name):
        read_tags_in_database()
        _, name, _, _ = image.unpacked_name
        assert name == expected_name


# endregion


class TestUpdateDatabase:
    # region tests

    def test_comprehensive_snapshot(self, snapshot, django_settings, elasticsearch, all_sources, tags):
        update_database()
        assert list(Card.objects.all().order_by("identifier")) == snapshot(name="cards")

    @pytest.mark.skip("we turned off upsert at time of writing because it's extremely slow with postgres")
    def test_upsert(self, django_settings, elasticsearch, all_sources):
        update_database()
        pk_to_identifier_1 = {x.pk: x.identifier for x in Card.objects.all()}
        update_database()
        pk_to_identifier_2 = {x.pk: x.identifier for x in Card.objects.all()}
        assert pk_to_identifier_1 == pk_to_identifier_2

    # endregion
