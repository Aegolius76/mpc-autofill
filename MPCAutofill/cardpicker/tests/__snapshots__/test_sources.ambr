# name: TestUpdateDatabase.test_comprehensive_snapshot[cards]
  list([
    <Card: [Example Drive 1] Plains [Type: CARD, Identifier: 1743iShpNEhY5xQ2PCUNpca4UL5t4rG3g, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] Delver of Secrets [Type: CARD, Identifier: 17fopRCNRge72U8Hac8pApHZtEalx5kHy, Uploaded: 09/10/2022, Priority: 2]>,
    <Card: [Example Drive 1] Huntmaster of the Fells [Type: CARD, Identifier: 1991MWCur9NdAFi-tQQD5YbQj2oqV_WRy, Uploaded: 22/09/2022, Priority: 2]>,
    <Card: [Example Drive 1] Brainstorm [Type: CARD, Identifier: 1c4M-sK9gd0Xju0NXCPtqeTW_DQTldVU5, Uploaded: 22/09/2022, Priority: 2]>,
    <Card: [Example Drive 1] Mountain [Type: CARD, Identifier: 1-dcs0FEE05MTGiYbKqs9HnRdhXkgtIJG, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 2] Past in Flames [Type: CARD, Identifier: 1dxSLHtw-VwwE09pZCA8OA6LbuWRZPEoU, Uploaded: 24/09/2022, Priority: 2]>,
    <Card: [Example Drive 1] Island (William Bradford) [Type: CARD, Identifier: 1HsvTYs1jFGe1c8U1PnNZ9aB8jkAW7KU0, Uploaded: 22/09/2022, Priority: 6]>,
    <Card: [Example Drive 1] Island [Type: CARD, Identifier: 1IDtqSjJ4Yo45AnNA4SplOiN7ewibifMa, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] Simple Cube [Type: CARDBACK, Identifier: 1JtXL6Ca9nQkvhwZZRR9ZuKA9_DzsFf1V, Uploaded: 22/09/2022, Priority: 17]>,
    <Card: [Example Drive 1] Ravager of the Fells [Type: CARD, Identifier: 1lv8WC1Xf1qxA7VHSc8jOtT5up6FwaBPH, Uploaded: 22/09/2022, Priority: 2]>,
    <Card: [Example Drive 1] Insectile Aberration [Type: CARD, Identifier: 1mO73GTYlieP0kiZEkF58pJSrZTmC9lNh, Uploaded: 09/10/2022, Priority: 2]>,
    <Card: [Example Drive 1] Swamp [Type: CARD, Identifier: 1O5Y_lBOBJK1D9Tw6pcsQJXYoTZXhR3mV, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] Simple Lotus [Type: CARDBACK, Identifier: 1oigI6wz0zA--pNMuExKTs40kBNH6VRP_, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] Forest [Type: CARD, Identifier: 1Sne_vn7TlohIbsjt933Kd5k2ff6vFS1I, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] PÃ¡st in Flames [Type: CARD, Identifier: 1UPdh7J7hScg4ZnxSPJ-EeBYHLp2s3Oz1, Uploaded: 24/09/2022, Priority: 2]>,
    <Card: [Example Drive 1] Goblin [Type: TOKEN, Identifier: 1V5E0avDmNyEUuFfYwx3nA05aj-1HY0rA, Uploaded: 09/10/2022, Priority: 2]>,
  ])
# ---
